{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import higlass\n",
    "from higlass.client import View, Track, CombinedTrack\n",
    "from higlass.tilesets import cooler, beddb, chromsizes, bigwig, Tileset\n",
    "import clodius\n",
    "import os\n",
    "import os.path as path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import negspy.coordinates as nc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = beddb(\"../2019-10-24_higlass/Data/hg38/gene-annotations-hg38.beddb\")\n",
    "chrom_sizes = chromsizes(\"../2019-10-24_higlass/hg38.chrom.sizes\")\n",
    "label_font_size = 18\n",
    "chr_label_size = 30\n",
    "annots_size = 150\n",
    "\n",
    "chrom_labels = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-chromosome-labels\",\n",
    "        tileset=chrom_sizes,\n",
    "        position=p,\n",
    "        height=chr_label_size,\n",
    "        width=chr_label_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "\n",
    "gene_annots = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-gene-annotations\",\n",
    "        tileset=genes,\n",
    "        position=p,\n",
    "        height=annots_size,\n",
    "        width=annots_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "\n",
    "hg38 = nc.get_chrominfo(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data\n",
    "## Contact matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "tumour_metadata = pd.read_csv(path.join(\"..\", \"..\", \"Data\", \"External\", \"LowC_Samples_Data_Available.tsv\"), sep=\"\\t\", header=0)\n",
    "tumour_metadata = tumour_metadata.loc[tumour_metadata.Include == \"Yes\", :]\n",
    "tumour_metadata[\"SampleID\"] = [\"PCa\" + str(i) for i in tumour_metadata[\"Sample ID\"]]\n",
    "\n",
    "tumour_samples = tumour_metadata[\"SampleID\"].tolist()\n",
    "t2e_samples = tumour_metadata.loc[tumour_metadata[\"T2E Status\"] == \"Yes\", \"SampleID\"].tolist()\n",
    "nont2e_samples = tumour_metadata.loc[tumour_metadata[\"T2E Status\"] == \"No\", \"SampleID\"].tolist()\n",
    "\n",
    "benign_metadata = pd.read_csv(path.join(\"..\", \"..\", \"Data\", \"Raw\", \"191220_A00827_0104_AHMW25DMXX_HiC\", \"config.tsv\"), sep=\"\\t\", header=0)\n",
    "benign_metadata = benign_metadata.loc[benign_metadata.Include == \"Yes\", :]\n",
    "benign_samples = benign_metadata[\"Sample\"].tolist()\n",
    "\n",
    "cell_line_metadata = pd.read_csv(path.join(\"..\", \"..\", \"Data\", \"External\", \"Rhie_2019\", \"config.tsv\"), sep=\"\\t\", header=0)\n",
    "cell_line_samples = cell_line_metadata[\"Run_Accession\"].tolist()\n",
    "\n",
    "all_samples = tumour_samples + benign_samples + cell_line_samples\n",
    "metadata = pd.read_csv(path.join(\"..\", \"2020-01-15_TAD-aggregation\", \"config.tsv\"), sep=\"\\t\", index_col=False, header=0)\n",
    "cooler_files = (\n",
    "    [path.join(\"..\", \"..\", \"Data\", \"Processed\", \"2019-06-18_PCa-LowC-sequencing\", \"Contacts\", s + \".mcool\") for s in tumour_samples + benign_samples]\n",
    "    + [path.join(\"..\", \"..\", \"Data\", \"External\", \"Rhie_2019\", \"Contacts\", s + \".mcool\") for s in cell_line_samples]\n",
    ")\n",
    "lowc_tilesets = {s: cooler(f) for s, f in zip(all_samples, cooler_files)}\n",
    "\n",
    "resolutions = [\n",
    "    1000, 2000, 3000, 4000, 5000,\n",
    "    10000, 20000, 30000, 40000, 50000,\n",
    "    100000, 200000, 300000, 400000, 500000,\n",
    "    1000000, 2000000, 3000000, 4000000, 5000000\n",
    "][::-1]\n",
    "\n",
    "min_resolution = 40000\n",
    "heatmap_size = 250\n",
    "colour_range = [\n",
    "    \"rgba(65, 105, 225, 1.0)\",\n",
    "    \"rgba(255, 250, 250, 1.0)\",\n",
    "    # \"rgba(240, 128, 128, 1.0)\",\n",
    "    \"rgba(255, 25, 25, 1.0)\"\n",
    "]\n",
    "lowc_heatmaps = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=l + \"heatmap\",\n",
    "            position=p,\n",
    "            tileset=lowc_tilesets[s],\n",
    "            filetype=\"cooler\",\n",
    "            height=heatmap_size,\n",
    "            options={\n",
    "                \"maxZoom\": str(resolutions.index(min_resolution)),\n",
    "                \"colorbarPosition\": \"topRight\",\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0],\n",
    "                \"colorRange\": colour_range,\n",
    "            },\n",
    "        ) for p, l in zip(\n",
    "            [\"top\", \"bottom\", \"left\", \"right\", \"center\"],\n",
    "            [\"horizontal-\", \"horizontal-\", \"vertical-\", \"vertical-\", \"\"]\n",
    "        )\n",
    "    } for s in all_samples\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## H3K27ac ChIP-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "line_height = 150\n",
    "chip_files = {s: path.join(\"..\", \"..\", \"Data\", \"Processed\", \"2019-05-03_PCa-H3K27ac-peaks\", \"Peaks\", s + \"_FE.sorted.filtered.bw\") for s in tumour_samples}\n",
    "chip_tilesets = {s: bigwig(f) for s, f in chip_files.items()}\n",
    "chip = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"-bar\",\n",
    "            position=p,\n",
    "            tileset=ts,\n",
    "            filetype=\"vector\",\n",
    "            height=line_height,\n",
    "            width=line_height,\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0] + \" H3K27ac\",\n",
    "                \"barFillColor\": metadata.loc[metadata[\"SampleID\"] == s, \"Sample_Colour\"].values[0],\n",
    "            },\n",
    "        ) for p, t in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "    } for s, ts in chip_tilesets.items()\n",
    "}\n",
    "\n",
    "chip_line = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"-line\",\n",
    "            position=p,\n",
    "            tileset=ts,\n",
    "            filetype=\"vector\",\n",
    "            height=line_height,\n",
    "            width=line_height,\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0] + \" H3K27ac\",\n",
    "                \"lineStrokeColor\": metadata.loc[metadata[\"SampleID\"] == s, \"Sample_Colour\"].values[0],\n",
    "            },\n",
    "        ) for p, t in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "    } for s, ts in chip_tilesets.items()\n",
    "}\n",
    "\n",
    "peak_files = {s: path.join(\"..\", \"..\", \"Data\", \"Processed\", \"2019-05-03_PCa-H3K27ac-peaks\", \"Peaks\", s + \"_peaks.filtered.beddb\") for s in tumour_samples}\n",
    "peak_tilesets = {s: beddb(f) for s, f in peak_files.items()}\n",
    "peaks = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"bedlike\",\n",
    "            position=p,\n",
    "            tileset=ts,\n",
    "            filetype=\"beddb\",\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0] + \" H3K27ac Peaks\",\n",
    "                \"fillColor\": metadata.loc[metadata[\"SampleID\"] == s, \"Sample_Colour\"].values[0],\n",
    "            },\n",
    "        ) for p, t in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"\", \"\", \"vertical-\", \"vertical-\"])\n",
    "    } for s, ts in peak_tilesets.items()\n",
    "}\n",
    "\n",
    "core_files = {s: path.join(\"..\", \"..\", \"Data\", \"Processed\", \"2019-05-03_PCa-H3K27ac-peaks\", \"COREs\", s + \".cores.beddb\") for s in tumour_samples}\n",
    "core_tilesets = {s: beddb(f) for s, f in core_files.items()}\n",
    "cores = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"bedlike\",\n",
    "            position=p,\n",
    "            tileset=ts,\n",
    "            filetype=\"beddb\",\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0] + \" COREs\",\n",
    "                \"fillColor\": metadata.loc[metadata[\"SampleID\"] == s, \"Sample_Colour\"].values[0],\n",
    "            },\n",
    "        ) for p, t in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"\", \"\", \"vertical-\", \"vertical-\"])\n",
    "    } for s, ts in core_tilesets.items()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load breakpoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "BREAK_DIR = path.join(\"..\", \"2020-02-19_chromoplexy\")\n",
    "breaks = pd.read_csv(\n",
    "    path.join(BREAK_DIR, \"Graphs\", \"sv-breakpoints.paired.tsv\"),\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    index_col=False,\n",
    ")\n",
    "\n",
    "break_files_1D = {s: path.join(BREAK_DIR, \"Tracks\", s + \".breaks.beddb\") for s in tumour_samples}\n",
    "break_tilesets_1D = {\n",
    "    s: beddb(\n",
    "        v,\n",
    "        name=metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0] + \" Breaks\"\n",
    "    ) for s, v in break_files_1D.items()\n",
    "}\n",
    "break_tracks_1D = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"bedlike\",\n",
    "            position=p,\n",
    "            tileset=break_tilesets_1D[s],\n",
    "            filetype=\"beddb\",\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": metadata.loc[metadata[\"SampleID\"] == s, \"Label\"].values[0],\n",
    "            },\n",
    "        ) for p, t in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"\", \"\", \"vertical-\", \"vertical-\"])\n",
    "    } for s in tumour_samples\n",
    "}\n",
    "\n",
    "disruption_tests = pd.read_csv(path.join(\"..\", \"2020-02-19_chromoplexy\", \"Graphs\", \"sv-disruption-tests.tsv\"), sep=\"\\t\")\n",
    "bp_singletons = pd.read_csv(path.join(\"..\", \"2020-02-19_chromoplexy\", \"Graphs\", \"sv-breakpoints.tsv\"), sep=\"\\t\")\n",
    "bp_pairs = pd.read_csv(path.join(\"..\", \"2020-02-19_chromoplexy\", \"Graphs\", \"sv-breakpoints.paired.tsv\"), sep=\"\\t\")\n",
    "disruption_tads = pd.read_csv(path.join(\"..\", \"2020-02-19_sv-disruption-TADs\", \"sv-disruption-tests.TADs.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structural variants and novel COREs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ZNF516 & PMEPA1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Acetylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1c26f5f348944170bbc63272c4fce26f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'GgV3b8uEQYSV9msPg0CGxA', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_s = \"PCa13848\"\n",
    "zoom_offset = 1e5\n",
    "\n",
    "pos = {\n",
    "    \"x\": [\n",
    "        nc.chr_pos_to_genome_pos(\"chr18\", 76200000, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chr18\", 76600000, hg38),\n",
    "    ],\n",
    "    \"y\": [\n",
    "        nc.chr_pos_to_genome_pos(\"chr20\", 57600000, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chr20\", 57800000, hg38),\n",
    "    ]\n",
    "}\n",
    "\n",
    "comb_tracks = {\n",
    "    \"chip\": CombinedTrack([chip[s][\"top\"] for s in tumour_samples if s != mut_s]),\n",
    "}\n",
    "\n",
    "# views\n",
    "v = [\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            break_tracks_1D[mut_s][\"top\"],\n",
    "            chip[mut_s][\"top\"],\n",
    "            cores[mut_s][\"top\"],\n",
    "            peaks[mut_s][\"top\"],\n",
    "            comb_tracks[\"chip\"],\n",
    "        ] + [\n",
    "            peaks[s][\"top\"] for s in tumour_samples if s != mut_s\n",
    "        ] + [\n",
    "            cores[s][\"top\"] for s in tumour_samples if s != mut_s\n",
    "        ],\n",
    "        initialXDomain=p,\n",
    "    ) for p in pos.values()\n",
    "]\n",
    "\n",
    "display, server, viewconf = higlass.display(\n",
    "    views=v,\n",
    "    zoom_syncs=[v],\n",
    "    #location_syncs=[v],\n",
    "    value_scale_syncs=[\n",
    "        [(v[0], chip[s][\"top\"]) for s in tumour_samples],\n",
    "        [(v[1], chip[s][\"top\"]) for s in tumour_samples],\n",
    "    ],\n",
    "    server_port=8192\n",
    ")\n",
    "\n",
    "display"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rearrangement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "838f7b181ec340fcbd1f80f499365f0e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'MUQkIluOStypHGjrtOoWZQ', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_s = \"PCa13848\"\n",
    "nonmut_s = \"PCa40507\"\n",
    "\n",
    "pos = {\n",
    "    \"x\": [\n",
    "        nc.chr_pos_to_genome_pos(\"chr18\", 76200000, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chr18\", 76600000, hg38),\n",
    "    ],\n",
    "    \"y\": [\n",
    "        nc.chr_pos_to_genome_pos(\"chr20\", 57600000, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chr20\", 57800000, hg38),\n",
    "    ]\n",
    "}\n",
    "\n",
    "# views\n",
    "v = [\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            chip[mut_s][\"top\"],\n",
    "            peaks[mut_s][\"top\"],\n",
    "            cores[mut_s][\"top\"],\n",
    "            break_tracks_1D[mut_s][\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            gene_annots[\"left\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            chip[mut_s][\"left\"],\n",
    "            peaks[mut_s][\"left\"],\n",
    "            cores[mut_s][\"left\"],\n",
    "            break_tracks_1D[mut_s][\"left\"],\n",
    "            lowc_heatmaps[mut_s][\"center\"],\n",
    "            #lowc_heatmaps[mut_s][\"bottom\"],\n",
    "        ],\n",
    "        initialXDomain=pos[\"x\"],\n",
    "        initialYDomain=pos[\"y\"],\n",
    "    ),\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            chip[nonmut_s][\"top\"],\n",
    "            peaks[nonmut_s][\"top\"],\n",
    "            cores[nonmut_s][\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            gene_annots[\"left\"],\n",
    "            chip[nonmut_s][\"left\"],\n",
    "            peaks[nonmut_s][\"left\"],\n",
    "            cores[nonmut_s][\"left\"],\n",
    "            lowc_heatmaps[nonmut_s][\"center\"],\n",
    "            #lowc_heatmaps[nonmut_s][\"bottom\"],\n",
    "        ],\n",
    "        initialXDomain=pos[\"x\"],\n",
    "        initialYDomain=pos[\"y\"],\n",
    "    ),\n",
    "]\n",
    "\n",
    "display, server, viewconf = higlass.display(\n",
    "    views=v,\n",
    "    zoom_syncs=[v],\n",
    "    location_syncs=[v],\n",
    "    value_scale_syncs=[\n",
    "        [(v[0], chip[mut_s][\"top\"]), (v[1], chip[nonmut_s][\"top\"])],\n",
    "        [(v[0], chip[mut_s][\"left\"]), (v[1], chip[nonmut_s][\"left\"])],\n",
    "        [(v[0], chip[mut_s][\"top\"]), (v[0], chip[mut_s][\"left\"])],\n",
    "        [(v[0], lowc_heatmaps[mut_s][\"center\"]), (v[1], lowc_heatmaps[nonmut_s][\"center\"])],\n",
    "        [(v[0], lowc_heatmaps[mut_s][\"bottom\"]), (v[1], lowc_heatmaps[nonmut_s][\"bottom\"])],\n",
    "    ],\n",
    "    server_port=8192\n",
    ")\n",
    "\n",
    "display"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COL1A1 & RNF144A"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rearrangement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6e387ac3256046fcb2c39c22a0c1f52e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'HYaM6TUdSy6RTeCnidmuHg', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_s = \"PCa40507\"\n",
    "nonmut_s = \"PCa13266\"\n",
    "zoom_offset = 1e5\n",
    "\n",
    "pos = {\n",
    "    \"x\": [\n",
    "        nc.chr_pos_to_genome_pos(\"chr17\", 50100000 - zoom_offset, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chr17\", 50400000 + zoom_offset, hg38),\n",
    "    ],\n",
    "    \"y\": [\n",
    "        [\n",
    "            nc.chr_pos_to_genome_pos(\"chr17\", 11300000 - zoom_offset, hg38),\n",
    "            nc.chr_pos_to_genome_pos(\"chr17\", 11800000 + zoom_offset, hg38),   \n",
    "        ],\n",
    "        [\n",
    "            nc.chr_pos_to_genome_pos(\"chr2\", 6726000 - zoom_offset, hg38),\n",
    "            nc.chr_pos_to_genome_pos(\"chr2\", 7064000 + zoom_offset, hg38),\n",
    "        ]\n",
    "    ]\n",
    "}\n",
    "\n",
    "# views\n",
    "v = [\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            chip[mut_s][\"top\"],\n",
    "            peaks[mut_s][\"top\"],\n",
    "            cores[mut_s][\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            gene_annots[\"left\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            chip[mut_s][\"left\"],\n",
    "            peaks[mut_s][\"left\"],\n",
    "            cores[mut_s][\"left\"],\n",
    "            lowc_heatmaps[mut_s][\"center\"],\n",
    "            #lowc_heatmaps[mut_s][\"bottom\"],\n",
    "        ],\n",
    "        initialXDomain=pos[\"x\"],\n",
    "        initialYDomain=pos[\"y\"][0],\n",
    "    ),\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            chip[nonmut_s][\"top\"],\n",
    "            peaks[nonmut_s][\"top\"],\n",
    "            cores[nonmut_s][\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            gene_annots[\"left\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            chip[nonmut_s][\"left\"],\n",
    "            peaks[nonmut_s][\"left\"],\n",
    "            cores[nonmut_s][\"left\"],\n",
    "            lowc_heatmaps[nonmut_s][\"center\"],\n",
    "            #lowc_heatmaps[nonmut_s][\"bottom\"],\n",
    "        ],\n",
    "        initialXDomain=pos[\"x\"],\n",
    "        initialYDomain=pos[\"y\"][0],\n",
    "    ),\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            chip[mut_s][\"top\"],\n",
    "            peaks[mut_s][\"top\"],\n",
    "            cores[mut_s][\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            gene_annots[\"left\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            chip[mut_s][\"left\"],\n",
    "            peaks[mut_s][\"left\"],\n",
    "            cores[mut_s][\"left\"],\n",
    "            lowc_heatmaps[mut_s][\"center\"],\n",
    "            #lowc_heatmaps[mut_s][\"bottom\"],\n",
    "        ],\n",
    "        initialXDomain=pos[\"x\"],\n",
    "        initialYDomain=pos[\"y\"][1],\n",
    "    ),\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            gene_annots[\"top\"],\n",
    "            chip[nonmut_s][\"top\"],\n",
    "            peaks[nonmut_s][\"top\"],\n",
    "            cores[nonmut_s][\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            gene_annots[\"left\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            chip[nonmut_s][\"left\"],\n",
    "            peaks[nonmut_s][\"left\"],\n",
    "            cores[nonmut_s][\"left\"],\n",
    "            lowc_heatmaps[nonmut_s][\"center\"],\n",
    "            #lowc_heatmaps[nonmut_s][\"bottom\"],\n",
    "        ],\n",
    "        initialXDomain=pos[\"x\"],\n",
    "        initialYDomain=pos[\"y\"][1],\n",
    "    ),\n",
    "]\n",
    "\n",
    "display, server, viewconf = higlass.display(\n",
    "    views=v,\n",
    "    zoom_syncs=[v],\n",
    "    location_syncs=[\n",
    "        v[0:2],\n",
    "        v[2:4]\n",
    "    ],\n",
    "    value_scale_syncs=[\n",
    "        [(v[0], chip[mut_s][\"top\"]), (v[1], chip[nonmut_s][\"top\"])],\n",
    "        [(v[0], chip[mut_s][\"left\"]), (v[1], chip[nonmut_s][\"left\"])],\n",
    "        [(v[2], chip[mut_s][\"top\"]), (v[3], chip[nonmut_s][\"top\"])],\n",
    "        [(v[2], chip[mut_s][\"left\"]), (v[3], chip[nonmut_s][\"left\"])],\n",
    "        [(v[0], chip[mut_s][\"top\"]), (v[0], chip[mut_s][\"left\"])],\n",
    "        [(v[2], chip[mut_s][\"top\"]), (v[2], chip[mut_s][\"left\"])],\n",
    "        [(v[0], lowc_heatmaps[mut_s][\"center\"]), (v[1], lowc_heatmaps[nonmut_s][\"center\"])],\n",
    "        [(v[0], lowc_heatmaps[mut_s][\"bottom\"]), (v[1], lowc_heatmaps[nonmut_s][\"bottom\"])],\n",
    "        [(v[2], lowc_heatmaps[mut_s][\"center\"]), (v[3], lowc_heatmaps[nonmut_s][\"center\"])],\n",
    "        [(v[2], lowc_heatmaps[mut_s][\"bottom\"]), (v[3], lowc_heatmaps[nonmut_s][\"bottom\"])],\n",
    "    ],\n",
    "    server_port=8192\n",
    ")\n",
    "\n",
    "display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
