{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0a7e3e2c",
   "metadata": {},
   "source": [
    "# Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c832d54e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.4.5\n"
     ]
    }
   ],
   "source": [
    "import higlass\n",
    "print(higlass.__version__)\n",
    "from higlass.client import View, Track, CombinedTrack\n",
    "from higlass.tilesets import cooler, beddb, chromsizes, bigwig, Tileset\n",
    "from higlass.viewer import save_b64_image_to_png\n",
    "from cooler import Cooler\n",
    "\n",
    "import clodius\n",
    "import os\n",
    "import os.path as path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import itertools\n",
    "import negspy.coordinates as nc\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def bed2ddb(filepath, uuid=None, **kwargs):\n",
    "    from clodius.tiles.bed2ddb import get_2d_tileset_info, get_2D_tiles\n",
    "    from clodius.tiles.utils import tiles_wrapper_2d\n",
    "\n",
    "    return Tileset(\n",
    "        uuid=uuid,\n",
    "        tileset_info=lambda: get_2d_tileset_info(filepath),\n",
    "        tiles=lambda tids: tiles_wrapper_2d(\n",
    "            tids, lambda z, x, y: get_2D_tiles(filepath, z, x, y)[(x, y)]\n",
    "        ),\n",
    "        **kwargs\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "581fcb30",
   "metadata": {},
   "source": [
    "# Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4be606d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_dir = path.join(\"..\", \"results\", \"2019-10-24_higlass\")\n",
    "genes = beddb(path.join(annot_dir, \"Data\", \"hg38\", \"gene-annotations-hg38.beddb\"))\n",
    "chrom_sizes = chromsizes(path.join(\"2019-10-24_higlass\", \"hg38.chrom.sizes\"))\n",
    "label_font_size = 18\n",
    "chr_label_size = 30\n",
    "annots_size = 150\n",
    "\n",
    "chrom_labels = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-chromosome-labels\",\n",
    "        tileset=chrom_sizes,\n",
    "        position=p,\n",
    "        height=chr_label_size,\n",
    "        width=chr_label_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip(\n",
    "        [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "        [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"]\n",
    "    )\n",
    "}\n",
    "chrom_grid = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-chromosome-grid\",\n",
    "        tileset=chrom_sizes,\n",
    "        position=p,\n",
    "        height=chr_label_size,\n",
    "        width=chr_label_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip(\n",
    "        [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "        [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "\n",
    "gene_annots = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-gene-annotations\",\n",
    "        tileset=genes,\n",
    "        position=p,\n",
    "        height=annots_size,\n",
    "        width=annots_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "\n",
    "hg38 = nc.get_chrominfo(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "912e259e",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b993baa",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "56a11502",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Include</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Sample_Colour</th>\n",
       "      <th>Type_Colour</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>T2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PCa13266</td>\n",
       "      <td>CPCG0268</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#7F3C8D</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PCa13848</td>\n",
       "      <td>CPCG0366</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#11A579</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCa14121</td>\n",
       "      <td>CPCG0255</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#3969AC</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCa19121</td>\n",
       "      <td>CPCG0258</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#F2B701</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PCa3023</td>\n",
       "      <td>CPCG0324</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#E73F74</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PCa33173</td>\n",
       "      <td>CPCG0246</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#80BA5A</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PCa40507</td>\n",
       "      <td>CPCG0342</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#E68310</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PCa51852</td>\n",
       "      <td>CPCG0336</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#008695</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PCa53687</td>\n",
       "      <td>CPCG0339</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#CF1C90</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PCa56413</td>\n",
       "      <td>CPCG0331</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PCa57294</td>\n",
       "      <td>CPCG0365</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#4B4B8F</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCa58215</td>\n",
       "      <td>CPCG0345</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#A5AA99</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Benign-Prostate-1648133</td>\n",
       "      <td>BP1648133</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Benign-Prostate-1595983</td>\n",
       "      <td>BP1595983</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Benign-Prostate-1664855</td>\n",
       "      <td>BP1664855</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Benign-Prostate-3320942</td>\n",
       "      <td>BP3320942</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Benign-Prostate-3330581</td>\n",
       "      <td>BP3330581</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR7702334</td>\n",
       "      <td>22Rv1 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR7702335</td>\n",
       "      <td>22Rv1 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR8446383</td>\n",
       "      <td>RWPE1 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FFBB78</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRR8446384</td>\n",
       "      <td>RWPE1 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FFBB78</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR8446385</td>\n",
       "      <td>C4-2B Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR8446386</td>\n",
       "      <td>C4-2B Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR8446387</td>\n",
       "      <td>C4-2B Rep 3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>4DNFI6HDY7WZ</td>\n",
       "      <td>H1-hESC Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Embryonic Stem Cell</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4DNFITH978XV</td>\n",
       "      <td>H1-hESC Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Embryonic Stem Cell</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>4DNFIT64Q7A3</td>\n",
       "      <td>HAP-1 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4DNFINSKEZND</td>\n",
       "      <td>HAP-1 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>4DNFIIV4M7TF</td>\n",
       "      <td>GM12878 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>4DNFIXVAKX9Q</td>\n",
       "      <td>GM12878 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample_ID          Label Include     Source       Type  \\\n",
       "0                  PCa13266       CPCG0268     Yes    Primary  Malignant   \n",
       "1                  PCa13848       CPCG0366     Yes    Primary  Malignant   \n",
       "2                  PCa14121       CPCG0255     Yes    Primary  Malignant   \n",
       "3                  PCa19121       CPCG0258     Yes    Primary  Malignant   \n",
       "4                   PCa3023       CPCG0324     Yes    Primary  Malignant   \n",
       "5                  PCa33173       CPCG0246     Yes    Primary  Malignant   \n",
       "6                  PCa40507       CPCG0342     Yes    Primary  Malignant   \n",
       "7                  PCa51852       CPCG0336     Yes    Primary  Malignant   \n",
       "8                  PCa53687       CPCG0339     Yes    Primary  Malignant   \n",
       "9                  PCa56413       CPCG0331     Yes    Primary  Malignant   \n",
       "10                 PCa57294       CPCG0365     Yes    Primary  Malignant   \n",
       "11                 PCa58215       CPCG0345     Yes    Primary  Malignant   \n",
       "12  Benign-Prostate-1648133      BP1648133     Yes    Primary     Benign   \n",
       "13  Benign-Prostate-1595983      BP1595983     Yes    Primary     Benign   \n",
       "14  Benign-Prostate-1664855      BP1664855     Yes    Primary     Benign   \n",
       "15  Benign-Prostate-3320942      BP3320942     Yes    Primary     Benign   \n",
       "16  Benign-Prostate-3330581      BP3330581     Yes    Primary     Benign   \n",
       "17               SRR7702334    22Rv1 Rep 1     Yes  Cell Line  Malignant   \n",
       "18               SRR7702335    22Rv1 Rep 2     Yes  Cell Line  Malignant   \n",
       "19               SRR8446383    RWPE1 Rep 1     Yes  Cell Line     Benign   \n",
       "20               SRR8446384    RWPE1 Rep 2     Yes  Cell Line     Benign   \n",
       "21               SRR8446385    C4-2B Rep 1     Yes  Cell Line  Malignant   \n",
       "22               SRR8446386    C4-2B Rep 2     Yes  Cell Line  Malignant   \n",
       "23               SRR8446387    C4-2B Rep 3     Yes  Cell Line  Malignant   \n",
       "24             4DNFI6HDY7WZ  H1-hESC Rep 1     Yes  Cell Line      Other   \n",
       "25             4DNFITH978XV  H1-hESC Rep 2     Yes  Cell Line      Other   \n",
       "26             4DNFIT64Q7A3    HAP-1 Rep 1     Yes  Cell Line      Other   \n",
       "27             4DNFINSKEZND    HAP-1 Rep 2     Yes  Cell Line      Other   \n",
       "28             4DNFIIV4M7TF  GM12878 Rep 1     Yes  Cell Line      Other   \n",
       "29             4DNFIXVAKX9Q  GM12878 Rep 2     Yes  Cell Line      Other   \n",
       "\n",
       "   Sample_Colour Type_Colour               Tissue  T2E  \n",
       "0        #7F3C8D     #1F77B4             Prostate   No  \n",
       "1        #11A579     #1F77B4             Prostate  Yes  \n",
       "2        #3969AC     #1F77B4             Prostate   No  \n",
       "3        #F2B701     #1F77B4             Prostate  Yes  \n",
       "4        #E73F74     #1F77B4             Prostate  Yes  \n",
       "5        #80BA5A     #1F77B4             Prostate   No  \n",
       "6        #E68310     #1F77B4             Prostate  Yes  \n",
       "7        #008695     #1F77B4             Prostate  Yes  \n",
       "8        #CF1C90     #1F77B4             Prostate   No  \n",
       "9        #F97B72     #1F77B4             Prostate  Yes  \n",
       "10       #4B4B8F     #1F77B4             Prostate   No  \n",
       "11       #A5AA99     #1F77B4             Prostate   No  \n",
       "12       #000000     #AEC7E8             Prostate   No  \n",
       "13       #000000     #AEC7E8             Prostate   No  \n",
       "14       #000000     #AEC7E8             Prostate   No  \n",
       "15       #000000     #AEC7E8             Prostate   No  \n",
       "16       #000000     #AEC7E8             Prostate   No  \n",
       "17       #000000     #FF7F0D             Prostate   No  \n",
       "18       #000000     #FF7F0D             Prostate   No  \n",
       "19       #000000     #FFBB78             Prostate   No  \n",
       "20       #000000     #FFBB78             Prostate   No  \n",
       "21       #000000     #FF7F0D             Prostate   No  \n",
       "22       #000000     #FF7F0D             Prostate   No  \n",
       "23       #000000     #FF7F0D             Prostate   No  \n",
       "24       #000000     #000000  Embryonic Stem Cell   No  \n",
       "25       #000000     #000000  Embryonic Stem Cell   No  \n",
       "26       #000000     #000000          Bone Marrow   No  \n",
       "27       #000000     #000000          Bone Marrow   No  \n",
       "28       #000000     #000000     Peripheral Blood   No  \n",
       "29       #000000     #000000     Peripheral Blood   No  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(\"config.tsv\", sep=\"\\t\")\n",
    "meta = meta.loc[meta.Include == \"Yes\", :]\n",
    "\n",
    "SAMPLES = {\n",
    "    \"all\": meta[\"Sample_ID\"].tolist(),\n",
    "    \"tumour\": meta.loc[(meta.Source == \"Primary\") & (meta.Type == \"Malignant\"), \"Sample_ID\"].tolist(),\n",
    "    \"benign\": meta.loc[(meta.Source == \"Primary\") & (meta.Type == \"Benign\"), \"Sample_ID\"].tolist(),\n",
    "    \"primary\": meta.loc[(meta.Source == \"Primary\"), \"Sample_ID\"].tolist(),\n",
    "    \"clines\": meta.loc[(meta.Source == \"Cell Line\"), \"Sample_ID\"].tolist(),\n",
    "}\n",
    "meta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94d65a86",
   "metadata": {},
   "source": [
    "## Contact matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "081b269e",
   "metadata": {},
   "outputs": [],
   "source": [
    "resolutions = [\n",
    "    1000, 2000, 3000, 4000, 5000,\n",
    "    10000, 20000, 30000, 40000, 50000,\n",
    "    100000, 200000, 300000, 400000, 500000,\n",
    "    1000000, 2000000, 3000000, 4000000, 5000000\n",
    "][::-1]\n",
    "min_resolution = 10000\n",
    "colour_range = [\n",
    "    \"rgba(65, 105, 225, 1.0)\",\n",
    "    \"rgba(255, 250, 250, 1.0)\",\n",
    "    \"rgba(255, 25, 25, 1.0)\"\n",
    "]\n",
    "\n",
    "mtx_dir = {\n",
    "    \"primary\": path.join(\"..\", \"Data\", \"Processed\", \"2019-06-18_PCa-LowC-sequencing\", \"Contacts\"),\n",
    "    \"clines\": path.join(\"..\", \"..\", \"Data\", \"External\", \"Rhie_2019\", \"Contacts\"),\n",
    "}\n",
    "mtx_files = {}\n",
    "for s in SAMPLES[\"primary\"]:\n",
    "    mtx_files[s] = path.join(mtx_dir[\"primary\"], s + \".mcool\")\n",
    "\n",
    "for s in SAMPLES[\"clines\"]:\n",
    "    mtx_files[s] = path.join(mtx_dir[\"clines\"], s + \".mcool\")\n",
    "\n",
    "mtx_tilesets = {\n",
    "    s: cooler(f)\n",
    "    for s, f in mtx_files.items()\n",
    "}\n",
    "\n",
    "contacts = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=pl + \"heatmap\",\n",
    "            datatype=\"matrix\",\n",
    "            filetype=\"cooler\",\n",
    "            tileset=mtx_tilesets[s],\n",
    "            height=height,\n",
    "            options={\n",
    "                \"maxZoom\": str(resolutions.index(min_resolution)),\n",
    "                \"colorbarPosition\": \"topRight\",\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0],\n",
    "                \"colorRange\": colour_range,\n",
    "            }\n",
    "        )\n",
    "        for p, pl, height in zip(\n",
    "            [\"top\", \"right\", \"bottom\", \"left\", \"center\"],\n",
    "            [\"\", \"\", \"\", \"\", \"\"],\n",
    "            #[\"horizontal-\", \"vertical-\", \"horizontal-\", \"vertical-\", \"\"],\n",
    "            #[\"linear-\", \"linear-\", \"linear-\", \"linear-\", \"\"],\n",
    "            [120, 120, 120, 120, 480],\n",
    "        )\n",
    "    }\n",
    "    for s in mtx_files.keys()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d477f52",
   "metadata": {},
   "source": [
    "## Structural variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bc9ab1a8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>chr_y</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>breakpoint_ID_x</th>\n",
       "      <th>breakpoint_ID_y</th>\n",
       "      <th>component_ID_x</th>\n",
       "      <th>component_ID_y</th>\n",
       "      <th>test_ID_x</th>\n",
       "      <th>test_ID_y</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>sv_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr3</td>\n",
       "      <td>73800000</td>\n",
       "      <td>75000000</td>\n",
       "      <td>chr4</td>\n",
       "      <td>179100000</td>\n",
       "      <td>179600000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>PCa13266</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr3</td>\n",
       "      <td>76010000</td>\n",
       "      <td>76130000</td>\n",
       "      <td>chr4</td>\n",
       "      <td>181680000</td>\n",
       "      <td>181710000</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>PCa13266</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>84300000</td>\n",
       "      <td>84600000</td>\n",
       "      <td>chr6</td>\n",
       "      <td>127700000</td>\n",
       "      <td>128300000</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>PCa13266</td>\n",
       "      <td>DUP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>75460000</td>\n",
       "      <td>75560000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>79350000</td>\n",
       "      <td>79420000</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>PCa13848</td>\n",
       "      <td>DUP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>151000000</td>\n",
       "      <td>151300000</td>\n",
       "      <td>chr13</td>\n",
       "      <td>99400000</td>\n",
       "      <td>100600000</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>31</td>\n",
       "      <td>28</td>\n",
       "      <td>PCa13848</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>chr12</td>\n",
       "      <td>108100000</td>\n",
       "      <td>108600000</td>\n",
       "      <td>chr7</td>\n",
       "      <td>113700000</td>\n",
       "      <td>116100000</td>\n",
       "      <td>308</td>\n",
       "      <td>309</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>308</td>\n",
       "      <td>309</td>\n",
       "      <td>PCa58215</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>chr12</td>\n",
       "      <td>128100000</td>\n",
       "      <td>130500000</td>\n",
       "      <td>chr7</td>\n",
       "      <td>106500000</td>\n",
       "      <td>108000000</td>\n",
       "      <td>315</td>\n",
       "      <td>311</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>311</td>\n",
       "      <td>312</td>\n",
       "      <td>PCa58215</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr12</td>\n",
       "      <td>128100000</td>\n",
       "      <td>130500000</td>\n",
       "      <td>chr7</td>\n",
       "      <td>138200000</td>\n",
       "      <td>139200000</td>\n",
       "      <td>315</td>\n",
       "      <td>310</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>311</td>\n",
       "      <td>310</td>\n",
       "      <td>PCa58215</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>chr13</td>\n",
       "      <td>30100000</td>\n",
       "      <td>30600000</td>\n",
       "      <td>chr13</td>\n",
       "      <td>73700000</td>\n",
       "      <td>75300000</td>\n",
       "      <td>313</td>\n",
       "      <td>314</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>314</td>\n",
       "      <td>315</td>\n",
       "      <td>PCa58215</td>\n",
       "      <td>DEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>chr3</td>\n",
       "      <td>77400000</td>\n",
       "      <td>79100000</td>\n",
       "      <td>chr4</td>\n",
       "      <td>26400000</td>\n",
       "      <td>27500000</td>\n",
       "      <td>307</td>\n",
       "      <td>306</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>307</td>\n",
       "      <td>306</td>\n",
       "      <td>PCa58215</td>\n",
       "      <td>BND</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>194 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr_x    start_x      end_x  chr_y    start_y      end_y  \\\n",
       "0     chr3   73800000   75000000   chr4  179100000  179600000   \n",
       "1     chr3   76010000   76130000   chr4  181680000  181710000   \n",
       "2     chr6   84300000   84600000   chr6  127700000  128300000   \n",
       "3     chr1   75460000   75560000   chr1   79350000   79420000   \n",
       "4     chr1  151000000  151300000  chr13   99400000  100600000   \n",
       "..     ...        ...        ...    ...        ...        ...   \n",
       "189  chr12  108100000  108600000   chr7  113700000  116100000   \n",
       "190  chr12  128100000  130500000   chr7  106500000  108000000   \n",
       "191  chr12  128100000  130500000   chr7  138200000  139200000   \n",
       "192  chr13   30100000   30600000  chr13   73700000   75300000   \n",
       "193   chr3   77400000   79100000   chr4   26400000   27500000   \n",
       "\n",
       "     breakpoint_ID_x  breakpoint_ID_y  component_ID_x  component_ID_y  \\\n",
       "0                  1                0               0               0   \n",
       "1                  3                2               1               1   \n",
       "2                  4                5               2               2   \n",
       "3                  6                7               0               0   \n",
       "4                 29               28              11              11   \n",
       "..               ...              ...             ...             ...   \n",
       "189              308              309               1               1   \n",
       "190              315              311               2               2   \n",
       "191              315              310               2               2   \n",
       "192              313              314               3               3   \n",
       "193              307              306               0               0   \n",
       "\n",
       "     test_ID_x  test_ID_y  SampleID sv_type  \n",
       "0            1          0  PCa13266     BND  \n",
       "1            3          2  PCa13266     BND  \n",
       "2            4          5  PCa13266     DUP  \n",
       "3            6          7  PCa13848     DUP  \n",
       "4           31         28  PCa13848     BND  \n",
       "..         ...        ...       ...     ...  \n",
       "189        308        309  PCa58215     BND  \n",
       "190        311        312  PCa58215     BND  \n",
       "191        311        310  PCa58215     BND  \n",
       "192        314        315  PCa58215     DEL  \n",
       "193        307        306  PCa58215     BND  \n",
       "\n",
       "[194 rows x 14 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_dir = path.join(\"..\", \"results\", \"2020-02-19_chromoplexy\")\n",
    "sv_track_dir = path.join(sv_dir, \"Tracks\")\n",
    "breaks = pd.read_csv(\n",
    "    path.join(sv_dir, \"Graphs\", \"sv-breakpoints.paired.tsv\"),\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    index_col=False,\n",
    ")\n",
    "\n",
    "sv_files = {\n",
    "    \"1D\": {\n",
    "        s: path.join(sv_track_dir, s + \".breaks.beddb\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: path.join(sv_track_dir, s + \".breaks.bed2ddb\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    }\n",
    "}\n",
    "sv_tilesets = {\n",
    "    \"1D\": {\n",
    "        s: beddb(\n",
    "            v,\n",
    "            name=meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Breaks\"\n",
    "        )\n",
    "        for s, v in sv_files[\"1D\"].items()\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: bed2ddb(\n",
    "            v,\n",
    "            name=meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Breaks\"\n",
    "        )\n",
    "        for s, v in sv_files[\"2D\"].items()\n",
    "    }\n",
    "}\n",
    "svs = {\n",
    "    \"1D\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=t + \"bedlike\",\n",
    "                datatype=\"bedlike\",\n",
    "                filetype=\"beddb\",\n",
    "                position=p,\n",
    "                tileset=sv_tilesets[\"1D\"][s],\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0],\n",
    "                    \"labelPosition\": \"topLeft\",\n",
    "                    \"labelColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"fillColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            ) for p, t in zip(\n",
    "                [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "                [\"\", \"\", \"vertical-\", \"vertical-\"])\n",
    "        }\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=pl+\"2d-rectangle-domains\",\n",
    "                datatype=\"2d-rectangle-domains\",\n",
    "                filetype=\"bed2ddb\",\n",
    "                position=p,\n",
    "                tileset=sv_tilesets[\"2D\"][s],\n",
    "                height=50,\n",
    "                width=50,\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta.Sample_ID == s, \"Label\"].values[0],\n",
    "                    \"labelPosition\": \"topLeft\",\n",
    "                    \"labelColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainFillColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            )\n",
    "            for p, pl in zip(\n",
    "                [\"top\", \"right\", \"bottom\", \"left\", \"center\"],\n",
    "                # later versions of HiGlass switch \"horizontal\"/\"vertical\" to \"linear\"\n",
    "                [\"horizontal-\", \"vertical-\", \"horizontal-\", \"vertical-\", \"\"],\n",
    "            )\n",
    "        }\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    }\n",
    "}\n",
    "breaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d4e693",
   "metadata": {},
   "source": [
    "# Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35f32aef",
   "metadata": {},
   "source": [
    "This will show the contact matrix where a single structural variant breakpoint is detected.\n",
    "Start by specifying the row index in the `breaks` table that you want to show."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d68e03aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_idx = 19"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9e53acb",
   "metadata": {},
   "source": [
    "The next cell will load the relevant contact matrix, breakpoints, and more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ec3a04eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "    chr_x   start_x     end_x  chr_y   start_y     end_y  breakpoint_ID_x  \\\n",
      "19  chr18  76340000  76460000  chr20  57700000  58220000               40   \n",
      "\n",
      "    breakpoint_ID_y  component_ID_x  component_ID_y  test_ID_x  test_ID_y  \\\n",
      "19               41              17              17         42         43   \n",
      "\n",
      "    SampleID sv_type  \n",
      "19  PCa13848     BND  \n",
      "PCa13848\n",
      "{'x': {'chr': 'chr18', 'start': 76240000.0, 'end': 76560000.0}, 'y': {'chr': 'chr20', 'start': 57600000.0, 'end': 58320000.0}}\n"
     ]
    }
   ],
   "source": [
    "# basepairs to show +/- around the SV\n",
    "offset = 1e5\n",
    "mut_sample = breaks.iloc[row_idx][\"SampleID\"]\n",
    "\n",
    "# get the component_ID of the breakpoint\n",
    "component_id = breaks.iloc[row_idx][\"component_ID_x\"]\n",
    "print(component_id)\n",
    "print(breaks.loc[\n",
    "    (breaks[\"SampleID\"] == mut_sample)\n",
    "    & (\n",
    "        (breaks[\"component_ID_x\"] == component_id)\n",
    "        | (breaks[\"component_ID_y\"] == component_id)\n",
    "    ),\n",
    "    :\n",
    "])\n",
    "\n",
    "\n",
    "# genomic coordinates\n",
    "coords = {\n",
    "    \"x\": {\n",
    "        \"chr\": breaks.iloc[row_idx][\"chr_x\"],\n",
    "        \"start\": breaks.iloc[row_idx][\"start_x\"] - offset,\n",
    "        \"end\": breaks.iloc[row_idx][\"end_x\"] + offset,\n",
    "    },\n",
    "    \"y\": {\n",
    "        \"chr\": breaks.iloc[row_idx][\"chr_y\"],\n",
    "        \"start\": breaks.iloc[row_idx][\"start_y\"] - offset,\n",
    "        \"end\": breaks.iloc[row_idx][\"end_y\"] + offset,\n",
    "    },\n",
    "}\n",
    "\n",
    "# convert to absolute plotting coordinates\n",
    "domains = {\n",
    "    \"x\": [\n",
    "        nc.chr_pos_to_genome_pos(coords[\"x\"][\"chr\"], coords[\"x\"][\"start\"], hg38),\n",
    "        nc.chr_pos_to_genome_pos(coords[\"x\"][\"chr\"], coords[\"x\"][\"end\"], hg38),\n",
    "    ],\n",
    "    \"y\": [\n",
    "        nc.chr_pos_to_genome_pos(coords[\"y\"][\"chr\"], coords[\"y\"][\"start\"], hg38),\n",
    "        nc.chr_pos_to_genome_pos(coords[\"y\"][\"chr\"], coords[\"y\"][\"end\"], hg38),\n",
    "    ],\n",
    "}\n",
    "print(mut_sample)\n",
    "print(coords)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a3c932",
   "metadata": {},
   "source": [
    "The cell below will show everything together in a combined plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f264fa25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f9414a2cbea341c3b931d26ca936db55",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'Oi58btfoR1Ky9FXIGJcniw', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [\n",
    "    View(\n",
    "        tracks=[\n",
    "            chrom_labels[\"top\"],\n",
    "            chrom_grid[\"top\"],\n",
    "            chrom_labels[\"left\"],\n",
    "            chrom_grid[\"left\"],\n",
    "            contacts[mut_sample][\"center\"],\n",
    "            svs[\"1D\"][mut_sample][\"top\"],\n",
    "            svs[\"1D\"][mut_sample][\"left\"],\n",
    "            svs[\"2D\"][mut_sample][\"left\"],\n",
    "        ],\n",
    "        initialXDomain=domains[\"x\"],\n",
    "        initialYDomain=domains[\"y\"],\n",
    "    ),\n",
    "]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d819cd3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
