{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import higlass\n",
    "from higlass.client import View, Track, CombinedTrack\n",
    "from higlass.tilesets import cooler, beddb, chromsizes, bigwig, Tileset\n",
    "from higlass.viewer import save_b64_image_to_png\n",
    "import clodius\n",
    "\n",
    "import os\n",
    "import os.path as path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import itertools\n",
    "import negspy.coordinates as nc\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def bed2ddb(filepath, uuid=None, **kwargs):\n",
    "    from clodius.tiles.bed2ddb import get_2d_tileset_info, get_2D_tiles\n",
    "    from clodius.tiles.utils import tiles_wrapper_2d\n",
    "\n",
    "    return Tileset(\n",
    "        uuid=uuid,\n",
    "        tileset_info=lambda: get_2d_tileset_info(filepath),\n",
    "        tiles=lambda tids: tiles_wrapper_2d(\n",
    "            tids, lambda z, x, y: get_2D_tiles(filepath, z, x, y)[(x, y)]\n",
    "        ),\n",
    "        **kwargs\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_dir = path.join(\"..\", \"results\", \"2019-10-24_higlass\")\n",
    "genes = beddb(path.join(annot_dir, \"data\", \"hg38\", \"gene-annotations-hg38.beddb\"))\n",
    "chrom_sizes = chromsizes(path.join(\"2019-10-24_higlass\", \"hg38.chrom.sizes\"))\n",
    "label_font_size = 18\n",
    "chr_label_size = 30\n",
    "annots_size = 150\n",
    "\n",
    "chrom_labels = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-chromosome-labels\",\n",
    "        tileset=chrom_sizes,\n",
    "        position=p,\n",
    "        height=chr_label_size,\n",
    "        width=chr_label_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "chrom_grid = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-chromosome-grid\",\n",
    "        tileset=chrom_sizes,\n",
    "        position=p,\n",
    "        height=chr_label_size,\n",
    "        width=chr_label_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip(\n",
    "        [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "        [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "\n",
    "gene_annots = {\n",
    "    p: Track(\n",
    "        track_type=l + \"-gene-annotations\",\n",
    "        tileset=genes,\n",
    "        position=p,\n",
    "        height=annots_size,\n",
    "        width=annots_size,\n",
    "        options={\n",
    "            \"fontSize\": label_font_size,\n",
    "            \"showMousePosition\": True,\n",
    "        },\n",
    "    ) for p, l in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "}\n",
    "\n",
    "hg38 = nc.get_chrominfo(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Include</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Sample_Colour</th>\n",
       "      <th>Type_Colour</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>T2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PCa13266</td>\n",
       "      <td>CPCG0268</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#7F3C8D</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PCa13848</td>\n",
       "      <td>CPCG0366</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#11A579</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCa14121</td>\n",
       "      <td>CPCG0255</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#3969AC</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCa19121</td>\n",
       "      <td>CPCG0258</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#F2B701</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PCa3023</td>\n",
       "      <td>CPCG0324</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#E73F74</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PCa33173</td>\n",
       "      <td>CPCG0246</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#80BA5A</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PCa40507</td>\n",
       "      <td>CPCG0342</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#E68310</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PCa51852</td>\n",
       "      <td>CPCG0336</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#008695</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PCa53687</td>\n",
       "      <td>CPCG0339</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#CF1C90</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PCa56413</td>\n",
       "      <td>CPCG0331</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PCa57294</td>\n",
       "      <td>CPCG0365</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#4B4B8F</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCa58215</td>\n",
       "      <td>CPCG0345</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#A5AA99</td>\n",
       "      <td>#1F77B4</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Benign-Prostate-1648133</td>\n",
       "      <td>BP1648133</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Benign-Prostate-1595983</td>\n",
       "      <td>BP1595983</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Benign-Prostate-1664855</td>\n",
       "      <td>BP1664855</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Benign-Prostate-3320942</td>\n",
       "      <td>BP3320942</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Benign-Prostate-3330581</td>\n",
       "      <td>BP3330581</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#AEC7E8</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR7702334</td>\n",
       "      <td>22Rv1 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR7702335</td>\n",
       "      <td>22Rv1 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR8446383</td>\n",
       "      <td>RWPE1 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FFBB78</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRR8446384</td>\n",
       "      <td>RWPE1 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Benign</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FFBB78</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR8446385</td>\n",
       "      <td>C4-2B Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR8446386</td>\n",
       "      <td>C4-2B Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR8446387</td>\n",
       "      <td>C4-2B Rep 3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#FF7F0D</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>4DNFI6HDY7WZ</td>\n",
       "      <td>H1-hESC Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Embryonic Stem Cell</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4DNFITH978XV</td>\n",
       "      <td>H1-hESC Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Embryonic Stem Cell</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>4DNFIT64Q7A3</td>\n",
       "      <td>HAP-1 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4DNFINSKEZND</td>\n",
       "      <td>HAP-1 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>4DNFIIV4M7TF</td>\n",
       "      <td>GM12878 Rep 1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>4DNFIXVAKX9Q</td>\n",
       "      <td>GM12878 Rep 2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Cell Line</td>\n",
       "      <td>Other</td>\n",
       "      <td>#000000</td>\n",
       "      <td>#000000</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample_ID          Label Include     Source       Type  \\\n",
       "0                  PCa13266       CPCG0268     Yes    Primary  Malignant   \n",
       "1                  PCa13848       CPCG0366     Yes    Primary  Malignant   \n",
       "2                  PCa14121       CPCG0255     Yes    Primary  Malignant   \n",
       "3                  PCa19121       CPCG0258     Yes    Primary  Malignant   \n",
       "4                   PCa3023       CPCG0324     Yes    Primary  Malignant   \n",
       "5                  PCa33173       CPCG0246     Yes    Primary  Malignant   \n",
       "6                  PCa40507       CPCG0342     Yes    Primary  Malignant   \n",
       "7                  PCa51852       CPCG0336     Yes    Primary  Malignant   \n",
       "8                  PCa53687       CPCG0339     Yes    Primary  Malignant   \n",
       "9                  PCa56413       CPCG0331     Yes    Primary  Malignant   \n",
       "10                 PCa57294       CPCG0365     Yes    Primary  Malignant   \n",
       "11                 PCa58215       CPCG0345     Yes    Primary  Malignant   \n",
       "12  Benign-Prostate-1648133      BP1648133     Yes    Primary     Benign   \n",
       "13  Benign-Prostate-1595983      BP1595983     Yes    Primary     Benign   \n",
       "14  Benign-Prostate-1664855      BP1664855     Yes    Primary     Benign   \n",
       "15  Benign-Prostate-3320942      BP3320942     Yes    Primary     Benign   \n",
       "16  Benign-Prostate-3330581      BP3330581     Yes    Primary     Benign   \n",
       "17               SRR7702334    22Rv1 Rep 1     Yes  Cell Line  Malignant   \n",
       "18               SRR7702335    22Rv1 Rep 2     Yes  Cell Line  Malignant   \n",
       "19               SRR8446383    RWPE1 Rep 1     Yes  Cell Line     Benign   \n",
       "20               SRR8446384    RWPE1 Rep 2     Yes  Cell Line     Benign   \n",
       "21               SRR8446385    C4-2B Rep 1     Yes  Cell Line  Malignant   \n",
       "22               SRR8446386    C4-2B Rep 2     Yes  Cell Line  Malignant   \n",
       "23               SRR8446387    C4-2B Rep 3     Yes  Cell Line  Malignant   \n",
       "24             4DNFI6HDY7WZ  H1-hESC Rep 1     Yes  Cell Line      Other   \n",
       "25             4DNFITH978XV  H1-hESC Rep 2     Yes  Cell Line      Other   \n",
       "26             4DNFIT64Q7A3    HAP-1 Rep 1     Yes  Cell Line      Other   \n",
       "27             4DNFINSKEZND    HAP-1 Rep 2     Yes  Cell Line      Other   \n",
       "28             4DNFIIV4M7TF  GM12878 Rep 1     Yes  Cell Line      Other   \n",
       "29             4DNFIXVAKX9Q  GM12878 Rep 2     Yes  Cell Line      Other   \n",
       "\n",
       "   Sample_Colour Type_Colour               Tissue  T2E  \n",
       "0        #7F3C8D     #1F77B4             Prostate   No  \n",
       "1        #11A579     #1F77B4             Prostate  Yes  \n",
       "2        #3969AC     #1F77B4             Prostate   No  \n",
       "3        #F2B701     #1F77B4             Prostate  Yes  \n",
       "4        #E73F74     #1F77B4             Prostate  Yes  \n",
       "5        #80BA5A     #1F77B4             Prostate   No  \n",
       "6        #E68310     #1F77B4             Prostate  Yes  \n",
       "7        #008695     #1F77B4             Prostate  Yes  \n",
       "8        #CF1C90     #1F77B4             Prostate   No  \n",
       "9        #F97B72     #1F77B4             Prostate  Yes  \n",
       "10       #4B4B8F     #1F77B4             Prostate   No  \n",
       "11       #A5AA99     #1F77B4             Prostate   No  \n",
       "12       #000000     #AEC7E8             Prostate   No  \n",
       "13       #000000     #AEC7E8             Prostate   No  \n",
       "14       #000000     #AEC7E8             Prostate   No  \n",
       "15       #000000     #AEC7E8             Prostate   No  \n",
       "16       #000000     #AEC7E8             Prostate   No  \n",
       "17       #000000     #FF7F0D             Prostate   No  \n",
       "18       #000000     #FF7F0D             Prostate   No  \n",
       "19       #000000     #FFBB78             Prostate   No  \n",
       "20       #000000     #FFBB78             Prostate   No  \n",
       "21       #000000     #FF7F0D             Prostate   No  \n",
       "22       #000000     #FF7F0D             Prostate   No  \n",
       "23       #000000     #FF7F0D             Prostate   No  \n",
       "24       #000000     #000000  Embryonic Stem Cell   No  \n",
       "25       #000000     #000000  Embryonic Stem Cell   No  \n",
       "26       #000000     #000000          Bone Marrow   No  \n",
       "27       #000000     #000000          Bone Marrow   No  \n",
       "28       #000000     #000000     Peripheral Blood   No  \n",
       "29       #000000     #000000     Peripheral Blood   No  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(\"config.tsv\", sep=\"\\t\")\n",
    "meta = meta.loc[meta.Include == \"Yes\", :]\n",
    "\n",
    "SAMPLES = {\n",
    "    \"all\": meta[\"Sample_ID\"].tolist(),\n",
    "    \"tumour\": meta.loc[(meta.Source == \"Primary\") & (meta.Type == \"Malignant\"), \"Sample_ID\"].tolist(),\n",
    "    \"benign\": meta.loc[(meta.Source == \"Primary\") & (meta.Type == \"Benign\"), \"Sample_ID\"].tolist(),\n",
    "    \"primary\": meta.loc[(meta.Source == \"Primary\"), \"Sample_ID\"].tolist(),\n",
    "    \"clines\": meta.loc[(meta.Source == \"Cell Line\"), \"Sample_ID\"].tolist(),\n",
    "}\n",
    "\n",
    "meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Contact Matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "resolutions = [\n",
    "    1000, 2000, 3000, 4000, 5000,\n",
    "    10000, 20000, 30000, 40000, 50000,\n",
    "    100000, 200000, 300000, 400000, 500000,\n",
    "    1000000, 2000000, 3000000, 4000000, 5000000\n",
    "][::-1]\n",
    "min_resolution = 40000\n",
    "colour_range = [\n",
    "    \"rgba(65, 105, 225, 1.0)\",\n",
    "    \"rgba(255, 250, 250, 1.0)\",\n",
    "    \"rgba(255, 25, 25, 1.0)\"\n",
    "]\n",
    "\n",
    "mtx_dir = {\n",
    "    \"primary\": path.join(\"..\", \"data\", \"Processed\", \"2019-06-18_PCa-LowC-sequencing\", \"Contacts\"),\n",
    "    \"clines\": path.join(\"..\", \"..\", \"data\", \"External\", \"Rhie_2019\", \"Contacts\"),\n",
    "}\n",
    "mtx_files = {}\n",
    "for s in SAMPLES[\"primary\"]:\n",
    "    mtx_files[s] = path.join(mtx_dir[\"primary\"], s + \".mcool\")\n",
    "\n",
    "for s in SAMPLES[\"clines\"]:\n",
    "    mtx_files[s] = path.join(mtx_dir[\"clines\"], s + \".mcool\")\n",
    "\n",
    "mtx_tilesets = {\n",
    "    s: cooler(f)\n",
    "    for s, f in mtx_files.items()\n",
    "}\n",
    "\n",
    "contacts = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=pl + \"heatmap\",\n",
    "            datatype=\"matrix\",\n",
    "            filetype=\"cooler\",\n",
    "            tileset=mtx_tilesets[s],\n",
    "            height=height,\n",
    "            options={\n",
    "                \"maxZoom\": str(resolutions.index(min_resolution)),\n",
    "                \"colorbarPosition\": \"topRight\",\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0],\n",
    "                \"colorRange\": colour_range,\n",
    "            }\n",
    "        )\n",
    "        for p, pl, height in zip(\n",
    "            [\"top\", \"right\", \"bottom\", \"left\", \"center\"],\n",
    "            [\"horizontal-\", \"horizontal-\", \"horizontal-\", \"horizontal-\", \"\"],\n",
    "            #[\"linear-\", \"linear-\", \"linear-\", \"linear-\", \"\"],\n",
    "            [120, 120, 120, 120, 240],\n",
    "        )\n",
    "    }\n",
    "    for s in mtx_files.keys()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b1f7181c7ccc49b08b78f7371a5e2148",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'YIS0dlGIRiS4OlRHgL5xAw', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "    ] + [\n",
    "        contacts[s][\"top\"] for s in SAMPLES[\"primary\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bea6493fc48b499485407e9044b6c396",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'V_8rOP1HTkW5H0i-89IeYQ', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "    ] + [\n",
    "        contacts[\"PCa13266\"][\"center\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr19\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chr20\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "windows = list(range(3, 21))\n",
    "tad_track_dir = path.join(\"..\", \"Results\", \"2020-08-29_TADs-downsampled\", \"Tracks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "tad_files = {\n",
    "    \"1D\": {\n",
    "        s: {\n",
    "            w: path.join(tad_track_dir, s + \".120000000.res_40000bp.domains.window_\" + str(w) + \".beddb\")\n",
    "            for w in windows\n",
    "        }\n",
    "        for s in SAMPLES[\"all\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: path.join(tad_track_dir, s + \".120000000.res_40000bp.domains.all.bed2ddb\")\n",
    "        for s in SAMPLES[\"all\"]\n",
    "    },\n",
    "}\n",
    "tad_tilesets = {\n",
    "    \"1D\": {\n",
    "        s: {\n",
    "            w: beddb(\n",
    "                tad_files[\"1D\"][s][w],\n",
    "                name=meta.loc[meta.Sample_ID == s, \"Label\"].values[0] + \"TADs (w = \" + str(w) +\")\"\n",
    "            )\n",
    "            for w in windows\n",
    "        }\n",
    "        for s in SAMPLES[\"all\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: bed2ddb(tad_files[\"2D\"][s])\n",
    "        for s in SAMPLES[\"all\"]\n",
    "    }\n",
    "}\n",
    "tads = {\n",
    "    \"1D\": {\n",
    "         s: {\n",
    "             w: {\n",
    "                p: Track(\n",
    "                    track_type=\"bedlike\",\n",
    "                    datatype=\"bedlike\",\n",
    "                    filetype=\"beddb\",\n",
    "                    position=p,\n",
    "                    tileset=tad_tilesets[\"1D\"][s][w],\n",
    "                    options={\n",
    "                        \"showMousePosition\": True,\n",
    "                        \"name\": meta.loc[meta.Sample_ID == s, \"Label\"].values[0],\n",
    "                        \"labelPosition\": \"topLeft\",\n",
    "                        \"fillColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                        \"labelColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0]\n",
    "                    },\n",
    "                )\n",
    "                for p in [\"top\", \"right\", \"bottom\", \"left\"]\n",
    "            }\n",
    "            for w in windows\n",
    "        }\n",
    "        for s in SAMPLES[\"all\"]   \n",
    "    },\n",
    "    \"2D\": {\n",
    "         s: {\n",
    "            p: Track(\n",
    "                track_type=pl+\"2d-rectangle-domains\",\n",
    "                datatype=\"2d-rectangle-domains\",\n",
    "                filetype=\"bed2ddb\",\n",
    "                position=p,\n",
    "                tileset=tad_tilesets[\"2D\"][s],\n",
    "                height=50,\n",
    "                width=50,\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta.Sample_ID == s, \"Label\"].values[0],\n",
    "                    \"labelPosition\": \"topLeft\",\n",
    "                    \"labelColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainFillColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainStrokeColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainOpacity\": 0.01,\n",
    "                },\n",
    "            )\n",
    "            for p, pl in zip(\n",
    "                [\"top\", \"right\", \"bottom\", \"left\", \"center\"],\n",
    "                # later versions of HiGlass switch \"horizontal\"/\"vertical\" to \"linear\"\n",
    "                [\"horizontal-\", \"vertical-\", \"horizontal-\", \"vertical-\", \"\"],\n",
    "            )\n",
    "        }\n",
    "        for s in SAMPLES[\"all\"]   \n",
    "    },\n",
    "}\n",
    "\n",
    "# a test track to show that 2D tracks are being displayed\n",
    "#ts = bed2ddb(path.join(tad_track_dir, \"tmp.bedpedb\"))\n",
    "#(d,s,v) = higlass.display([\n",
    "#    View([\n",
    "#        Track('top-axis'),\n",
    "#        Track('2d-rectangle-domains', position='center', tileset=ts, height=400),\n",
    "#    ], initialXDomain=[0, 20e6])\n",
    "#])\n",
    "#d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8157845ac2b9450f8d0265aec3a34898",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'eEqK58RsQHm_nEU42kjQ7A', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "    ] + [\n",
    "        tads[\"2D\"][s][\"top\"] for s in SAMPLES[\"primary\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compartments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmpt_files = {\n",
    "    s: path.join(\"..\", \"Results\", \"2021-03-03_compartments\", \"Compartments\", s + \".compartments.cis.bw\")\n",
    "    for s in SAMPLES[\"all\"]\n",
    "}\n",
    "cmpt_tilesets = {s: bigwig(f) for s, f in cmpt_files.items()}\n",
    "compartments = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"-line\",\n",
    "            filetype=\"vector\",\n",
    "            position=p,\n",
    "            tileset=ts,\n",
    "            height=100,\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Compartment\",\n",
    "                \"lineStrokeColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                \"valueScaleMin\": -2,\n",
    "                \"valueScaleMax\": 2,\n",
    "                \"zeroLineVisible\": True,\n",
    "                \"zeroLineColor\": \"#000000\",\n",
    "            },\n",
    "        ) for p, t in zip(\n",
    "            [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "            [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"]\n",
    "        )\n",
    "    } for s, ts in cmpt_tilesets.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "adc52efd91214f8aa48895eb5e127540",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'H92uDww_QtOC83Sd0tJAUw', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "    ] + [\n",
    "        compartments[s][\"top\"] for s in SAMPLES[\"primary\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    "    value_scale_syncs=[\n",
    "        [(views[0], compartments[s][\"top\"]) for s in SAMPLES[\"primary\"]],\n",
    "    ],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chromatin Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "loop_track_dir = path.join(\"..\", \"Results\", \"2020-10-06_loops\", \"Tracks\")\n",
    "loop_files = {\n",
    "    \"1D\": {\n",
    "        s: path.join(loop_track_dir, s + \".anchors.beddb\")\n",
    "        for s in SAMPLES[\"primary\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: path.join(loop_track_dir, s + \".loops.bed2ddb\")\n",
    "        for s in SAMPLES[\"primary\"]\n",
    "    },\n",
    "}\n",
    "loop_tilesets = {\n",
    "    \"1D\": {\n",
    "        s: beddb(f)\n",
    "        for s, f in loop_files[\"1D\"].items()\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: bed2ddb(f)\n",
    "        for s, f in loop_files[\"2D\"].items()\n",
    "    },\n",
    "}\n",
    "loops = {\n",
    "    \"1D\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=\"bedlike\",\n",
    "                datatype=\"bedlike\",\n",
    "                filetype=\"beddb\",\n",
    "                tileset=loop_tilesets[\"1D\"][s],\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Interactions\",\n",
    "                    \"lineStrokeColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                }\n",
    "            )\n",
    "            for p in [\"top\", \"left\", \"bottom\", \"right\"]\n",
    "        }\n",
    "        for s in SAMPLES[\"primary\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=\"2d-rectangle-domains\",\n",
    "                datatype=\"2d-rectangle-domains\",\n",
    "                filetype=\"bed2ddb\",\n",
    "                tileset=loop_tilesets[\"2D\"][s],\n",
    "                height=height,\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Interactions\",\n",
    "                    \"lineStrokeColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainStrokeColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainFillColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainOpcaity\": 0.4,\n",
    "                }\n",
    "            )\n",
    "            for p, height in zip(\n",
    "                [\"top\", \"left\", \"bottom\", \"right\", \"center\"],\n",
    "                [50, 50, 50, 50, 600],\n",
    "            )\n",
    "        }\n",
    "        for s in SAMPLES[\"primary\"]\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "26e6edecc9dc43eca729b3e07f5cb594",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'MjeaEw1VQ3agvtdbSmLnig', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "        chrom_labels[\"left\"],\n",
    "        gene_annots[\"left\"],\n",
    "        CombinedTrack([ loops[\"2D\"][s][\"center\"] for s in SAMPLES[\"primary\"] ])\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    server_port=8192,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Structural variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_dir = path.join(\"..\", \"Results\", \"2020-02-19_chromoplexy\")\n",
    "sv_track_dir = path.join(sv_dir, \"Tracks\")\n",
    "breaks = pd.read_csv(\n",
    "    path.join(sv_dir, \"Graphs\", \"sv-breakpoints.paired.tsv\"),\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    index_col=False,\n",
    ")\n",
    "\n",
    "sv_files = {\n",
    "    \"1D\": {\n",
    "        s: path.join(sv_track_dir, s + \".breaks.beddb\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: path.join(sv_track_dir, s + \".breaks.bed2ddb\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    }\n",
    "}\n",
    "sv_tilesets = {\n",
    "    \"1D\": {\n",
    "        s: beddb(\n",
    "            v,\n",
    "            name=meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Breaks\"\n",
    "        )\n",
    "        for s, v in sv_files[\"1D\"].items()\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: bed2ddb(\n",
    "            v,\n",
    "            name=meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" Breaks\"\n",
    "        )\n",
    "        for s, v in sv_files[\"2D\"].items()\n",
    "    }\n",
    "}\n",
    "svs = {\n",
    "    \"1D\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=t + \"bedlike\",\n",
    "                datatype=\"bedlike\",\n",
    "                filetype=\"beddb\",\n",
    "                position=p,\n",
    "                tileset=sv_tilesets[\"1D\"][s],\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0],\n",
    "                    \"labelPosition\": \"topLeft\",\n",
    "                    \"labelColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"fillColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            ) for p, t in zip(\n",
    "                [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "                [\"\", \"\", \"vertical-\", \"vertical-\"])\n",
    "        }\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "    \"2D\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=pl+\"2d-rectangle-domains\",\n",
    "                datatype=\"2d-rectangle-domains\",\n",
    "                filetype=\"bed2ddb\",\n",
    "                position=p,\n",
    "                tileset=sv_tilesets[\"2D\"][s],\n",
    "                height=50,\n",
    "                width=50,\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta.Sample_ID == s, \"Label\"].values[0],\n",
    "                    \"labelPosition\": \"topLeft\",\n",
    "                    \"labelColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                    \"rectangleDomainFillColor\": meta.loc[meta.Sample_ID == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            )\n",
    "            for p, pl in zip(\n",
    "                [\"top\", \"right\", \"bottom\", \"left\", \"center\"],\n",
    "                # later versions of HiGlass switch \"horizontal\"/\"vertical\" to \"linear\"\n",
    "                [\"horizontal-\", \"vertical-\", \"horizontal-\", \"vertical-\", \"\"],\n",
    "            )\n",
    "        }\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c19170e9e30b4b2ba12133f938f037ef",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'ARszapOTQ3OkOyxM8XZDVg', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "    ] + [\n",
    "        svs[\"1D\"][s][\"top\"] for s in SAMPLES[\"tumour\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    server_port=8193,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copy Number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_track_dir = path.join(\"..\", \"Data\", \"Processed\", \"2019-06-18_PCa-LowC-sequencing\", \"Tracks\")\n",
    "cnv_files = {\n",
    "    s: path.join(cnv_track_dir, s + \".cnv.bw\")\n",
    "    for s in SAMPLES[\"all\"]\n",
    "}\n",
    "cnv_tilesets = {\n",
    "    s: bigwig(f)\n",
    "    for s, f in cnv_files.items()\n",
    "}\n",
    "cnvs = {\n",
    "    s: {\n",
    "        p: Track(\n",
    "            track_type=t + \"-line\",\n",
    "            datatype=\"vector\",\n",
    "            filetype=\"vector\",\n",
    "            position=p,\n",
    "            tileset=ts,\n",
    "            height=100,\n",
    "            options={\n",
    "                \"showMousePosition\": True,\n",
    "                \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" log2CopyRatio\",\n",
    "                \"lineStrokeColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                \"valueScaleMin\": -2,\n",
    "                \"valueScaleMax\": 2,\n",
    "            },\n",
    "        ) for p, t in zip([\"top\", \"bottom\", \"left\", \"right\"], [\"horizontal\", \"horizontal\", \"vertical\", \"vertical\"])\n",
    "    } for s, ts in cnv_tilesets.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "98b013bcaf3a47229e3efe31b853cb57",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'NEOkkVRuRnCGmpGIOVqxUA', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        chrom_grid[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "    ] + [\n",
    "        cnvs[s][\"top\"] for s in SAMPLES[\"primary\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    server_port=8193,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    "    value_scale_syncs=[\n",
    "        [(views[0], cnvs[s][\"top\"]) for s in SAMPLES[\"primary\"]],\n",
    "    ],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## H3K27ac ChIP-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "chip_track_dir = path.join(\"..\", \"Data\", \"Processed\", \"2019-05-03_PCa-H3K27ac-peaks\", \"Tracks\")\n",
    "\n",
    "chip_files = {\n",
    "    \"signal\": {\n",
    "        s: path.join(chip_track_dir, \"Pca\" + s[3:] + \"_FE.sorted.filtered.bw\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "    \"peaks\": {\n",
    "        s: path.join(chip_track_dir, \"Pca\" + s[3:] + \"_peaks.beddb\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "    \"cores\": {\n",
    "        s: path.join(chip_track_dir, s + \".cores.beddb\")\n",
    "        for s in SAMPLES[\"tumour\"]\n",
    "    },\n",
    "}\n",
    "    \n",
    "chip_tilesets = {\n",
    "    \"signal\": {\n",
    "        s: bigwig(f)\n",
    "        for s, f in chip_files[\"signal\"].items()\n",
    "    },\n",
    "    \"peaks\": {\n",
    "        s: beddb(f)\n",
    "        for s, f in chip_files[\"peaks\"].items()\n",
    "    },\n",
    "    \"cores\": {\n",
    "        s: beddb(f)\n",
    "        for s, f in chip_files[\"cores\"].items()\n",
    "    },\n",
    "}\n",
    "\n",
    "line_height = 150\n",
    "\n",
    "chip = {\n",
    "    \"signal\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=t + \"line\",\n",
    "                position=p,\n",
    "                tileset=ts,\n",
    "                filetype=\"vector\",\n",
    "                height=line_height,\n",
    "                width=line_height,\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" H3K27ac\",\n",
    "                    \"lineStrokeColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            )\n",
    "            for p, t in zip(\n",
    "                [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "                [\"horizontal-\", \"horizontal-\", \"vertical-\", \"vertical-\"],\n",
    "            )\n",
    "        }\n",
    "        for s, ts in chip_tilesets[\"signal\"].items()\n",
    "    },\n",
    "    \"peaks\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=t + \"bedlike\",\n",
    "                position=p,\n",
    "                tileset=ts,\n",
    "                filetype=\"beddb\",\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" H3K27ac\",\n",
    "                    \"fillColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            )\n",
    "            for p, t in zip(\n",
    "                [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "                [\"\", \"\", \"vertical-\", \"vertical-\"],\n",
    "            )\n",
    "        }\n",
    "        for s, ts in chip_tilesets[\"peaks\"].items()\n",
    "    },\n",
    "    \"cores\": {\n",
    "        s: {\n",
    "            p: Track(\n",
    "                track_type=t + \"bedlike\",\n",
    "                position=p,\n",
    "                tileset=ts,\n",
    "                filetype=\"beddb\",\n",
    "                options={\n",
    "                    \"showMousePosition\": True,\n",
    "                    \"name\": meta.loc[meta[\"Sample_ID\"] == s, \"Label\"].values[0] + \" H3K27ac\",\n",
    "                    \"fillColor\": meta.loc[meta[\"Sample_ID\"] == s, \"Sample_Colour\"].values[0],\n",
    "                },\n",
    "            )\n",
    "            for p, t in zip(\n",
    "                [\"top\", \"bottom\", \"left\", \"right\"],\n",
    "                [\"\", \"\", \"vertical-\", \"vertical-\"],\n",
    "            )\n",
    "        }\n",
    "        for s, ts in chip_tilesets[\"cores\"].items()\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "801efe65a17d4ec989df11f2bea72409",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HiGlassDisplay(viewconf={'editable': True, 'views': [{'uid': 'ED-4tgSMTzenAluXhncuOw', 'tracks': {'top': [{'ty…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views = [View(\n",
    "    tracks=[\n",
    "        chrom_labels[\"top\"],\n",
    "        chrom_grid[\"top\"],\n",
    "        gene_annots[\"top\"],\n",
    "        #CombinedTrack([ chip[\"signal\"][s][\"top\"] for s in SAMPLES[\"tumour\"] ]),\n",
    "    ] + [\n",
    "        chip[\"signal\"][s][\"top\"] for s in SAMPLES[\"tumour\"]\n",
    "    ],\n",
    "    initialXDomain=[\n",
    "        nc.chr_pos_to_genome_pos(\"chr1\", 0, hg38),\n",
    "        nc.chr_pos_to_genome_pos(\"chrM\", 0, hg38)\n",
    "    ],\n",
    ")]\n",
    "\n",
    "d, s, v = higlass.display(\n",
    "    views=views,\n",
    "    server_port=8193,\n",
    "    zoom_syncs=[views],\n",
    "    location_syncs=[views],\n",
    "    value_scale_syncs=[\n",
    "        [(views[0], chip[\"signal\"][s][\"top\"]) for s in SAMPLES[\"tumour\"]],\n",
    "    ],\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
