# ==============================================================================
# Meta Rules
# ==============================================================================
rule all:
    input:
        "gencode.v33.genes.sorted.bed",
        "gencode.v33.tss.sorted.bed",
        "gencode.v33.all-genes.bed",
        "gencode.v33.all-genes.promoters.bed",
        "gencode.v33.all-transcripts.bed",

# ==============================================================================
# Rules
# ==============================================================================
rule download_annotation:
    output:
        temp("gencode.v33.annotation.gtf.gz")
    shell:
        "curl -O ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_33/gencode.v33.annotation.gtf.gz"

rule all_genes:
    input:
        script = "gene.awk",
        annot = "gencode.v33.annotation.gtf",
    output:
        "gencode.v33.all-genes.bed"
    shell:
        # use either tab or "; " as field separators, only keep genes
        "awk -f {input.script} {input.annot} > {output}"

rule all_transcripts:
    input:
        script = "transcript.awk",
        annot = "gencode.v33.annotation.gtf",
    output:
        "gencode.v33.all-transcripts.bed"
    shell:
        # use either tab or "; " as field separators, only keep transcripts
        "awk -f {input.script} {input.annot} > {output}"

rule promoters:
    input:
        script = "promoters.awk",
        annot = "gencode.v33.all-genes.bed",
    output:
        "gencode.v33.all-genes.promoters.bed",
    shell:
        "awk -f {input.script} {input.annot} > {output}"

rule filter:
    input:
        script = "filter-gencode.R",
        data = "gencode.v33.annotation.gtf"
    output:
        "gencode.v33.genes.bed",
        "gencode.v33.tss.bed",
    shell:
        "Rscript {input.script}"

rule sort:
    input:
        "{file}.bed"
    output:
        "{file}.sorted.bed"
    shell:
        "sort -k 1,1 -V -k2,2n {input} > {output}"
