# ==============================================================================
# Environment
# ==============================================================================
import os.path as path
import pandas as pd

CONFIG = pd.read_csv(
    path.join("..", "..", "Data", "External", "LowC_Samples_Data_Available.tsv"),
    sep="\t",
    header=0
)
SAMPLES = ["PCa" + str(i) for i in CONFIG.loc[CONFIG.Include == "Yes", "Sample ID"]]

GRAPH_DIR = path.join("..", "2020-02-19_chromoplexy", "Graphs")
PLOT_DIR = "Plots"
TAD_DIR = path.join("..", "2020-01-15_TAD-aggregation", "resolved-TADs")

PLOT_EXTS = ["png", "pdf"]

# ==============================================================================
# Meta Rules
# ==============================================================================
rule all:
    input:
        expand(
            path.join(PLOT_DIR, "expression.{plot}.{ext}"),
            plot=[
                "p-values",
                "qq",
                "z",
                "fold-change",
                "fold-change.thresholded",
                "fold-change-vs-difference",
                "fold-change.ecdf",
                "fold-change.ecdf.thresholded",
            ],
            ext=PLOT_EXTS,
        ),
        "sv-disruption-tests.expression.tsv",
        "sv-disruption-tests.expression.gene-level.tsv",
        "sv-disruption.expression.gene-level.stratified.all.tsv",
        "sv-disruption.expression.gene-level.stratified.thresholded.tsv",


# ==============================================================================
# Rules
# ==============================================================================
rule altered_expression:
    input:
        script = "altered-expression.py",
        graph = path.join(GRAPH_DIR, "breakpoints.all-samples.p"),
        rna = path.join("..", "..", "Data", "External", "CPC-GENE", "CPC-GENE_Chen-2019_RNAseq_rsem_gene_FPKM.13-LowC-only.tsv"),
        gencode = path.join("..", "..", "Data", "External", "GENCODE", "gencode.v33.all-genes.bed"),
        tads = expand(
            path.join(TAD_DIR, "separated-TADs", "{sample}.40000bp.w_3.domains.bed"),
            sample=SAMPLES
        ),
    output:
        "sv-disruption-tests.expression.tsv",
        "sv-disruption-tests.expression.gene-level.tsv",
    shell:
        "python {input.script}"

rule plot_expression:
    input:
        script = "plot-dge.R",
        rna = path.join("..", "..", "Data", "External", "CPC-GENE", "CPC-GENE_Chen-2019_RNAseq_rsem_gene_FPKM.13-LowC-only.tsv"),
        gencode = path.join("..", "..", "Data", "External", "GENCODE", "gencode.v33.all-genes.bed"),
        tests = "sv-disruption-tests.expression.tsv",
        genes = "sv-disruption-tests.expression.gene-level.tsv",
        chroms = path.join("..", "..", "Data", "Processed", "2019-06-18_PCa-LowC-sequencing", "hg38.sizes.txt"),
    output:
        expand(
            path.join(PLOT_DIR, "expression.{plot}.{ext}"),
            plot=[
                "p-values",
                "qq",
                "z",
                "fold-change",
                "fold-change.thresholded",
                "fold-change-vs-difference",
                "fold-change.ecdf",
                "fold-change.ecdf.thresholded",
            ],
            ext=PLOT_EXTS,
        ),
        "sv-disruption.expression.gene-level.stratified.all.tsv",
        "sv-disruption.expression.gene-level.stratified.thresholded.tsv",
    shell:
        "Rscript {input.script}"
